sns_ik_lib/src/sns_ik.cpp

```diff
@@ -362,6 +373,230 @@ bool SNS_IK::nullspaceBiasTask(const KDL::JntArray& q_bias,
+{
+  // Get the Jacobian corresponding to current joint position
+  KDL::Jacobian jacobian;
+  jacobian.resize(jnt_pos_in.rows());
+  if (m_jacobianSolver->JntToJac(jnt_pos_in, jacobian) < 0)
+  {
+    ROS_ERROR("JntToJac failed!");
+    return false;
+  }
+
+  *jacobianOut = jacobian.data;
+
+  return true;
+}
+
+/**
+ * Method to return the jacobian dot matrix.
+ *
+ * @param jnt_pos_in: vector of joint positions
+ * @param jnt_vel_in: vector of joint velocities
+ * @param jacobianDotOut: the jacobian dot matrix for the given joint angles and velocities
+ * @param tipName: std string holding the tip name.
+ *
+ * @return true: if jacobianDotOut is successfully loaded with the desired matrix.
+ *
+ */
+bool SNS_IK::getJacobianDot(const KDL::JntArray& jnt_pos_in, const KDL::JntArray& jnt_vel_in,
+                                  Eigen::MatrixXd *jacobianDotOut, const std::string & tipName)
+{
+  // copy joint positions and velocities into kdl JntArray
+  size_t numJnt = jnt_pos_in.rows();
+
+  // Get the Jacobian corresponding to current joint position
+  KDL::Jacobian jacobian;
+  jacobian.resize(jnt_pos_in.rows());
+  if (m_jacobianSolver->JntToJac(jnt_pos_in, jacobian) < 0)
+  {
+    ROS_ERROR("JntToJac failed!");
+    return false;
+  }
+
+  // Get Linear/Angular Jacobians
+  Eigen::MatrixXd Jv, Jw;
+  Jv = jacobian.data.block(0,0,3,numJnt);
+  Jw = jacobian.data.block(3,0,3,numJnt);
+
+  /// Compute Jacobian Dot
+  KDL::Jacobian jacobianDot;
+  jacobianDot.resize(numJnt);
+
+  // initialization
+  Eigen::MatrixXd p_0_i_1_i = Eigen::MatrixXd::Zero(3, numJnt);
+  Eigen::MatrixXd pdot_0_i_1_i = Eigen::MatrixXd::Zero(3, numJnt);
+  Eigen::MatrixXd p_0_i_1 = Eigen::MatrixXd::Zero(3, numJnt);
+  Eigen::MatrixXd p_0_i_1_k = Eigen::MatrixXd::Zero(3, numJnt);
+  Eigen::MatrixXd pdot_0_i_1_k = Eigen::MatrixXd::Zero(3, numJnt);
+
+  Eigen::MatrixXd w = Eigen::MatrixXd::Zero(3, numJnt);
+
+  Eigen::MatrixXd Jdw = Eigen::MatrixXd::Zero(3, numJnt);
+  Eigen::MatrixXd Jdv = Eigen::MatrixXd::Zero(3, numJnt);
+
+  // get FK of the intermediate links
+  std::vector<KDL::Frame> transforms;
+  KDL::ChainFkSolverPos_recursive fwdKin(m_chain);
+  KDL::JntArray q(jnt_pos_in.rows());
+  q.data << 1,1,1,1,1,1,1;
+
+  for (size_t i = 0; i < 9; i++) {
+    KDL::Frame transformTemp;
+    fwdKin.JntToCart(q, transformTemp, i+1);
+    if (i < numJnt)
+      p_0_i_1.col(i) << transformTemp.p(0),transformTemp.p(1),transformTemp.p(2);
+    transforms.push_back(transformTemp);
+  }
+
+  // get the tip position vector
+  Eigen::Vector3d p_0_k(transforms[8].p(0),transforms[8].p(1),transforms[8].p(2));
+
+  // first pass -- compute Jdw and quantities needed for Jdv
+  Eigen::Vector3d w_i(0,0,0);
+  Eigen::Vector3d w_i_1(0,0,0);
+  for (size_t i = 0; i < numJnt; i++) {
+    if (i > 0) {
+      w_i_1 = w.col(i-1);
+    }
+
+    // compute w_i and w_i_1
+    w_i = jnt_vel_in(i)*Jw.col(i) + w_i_1;
+    w.col(i) = w_i;
+
+    // compute Jdw
+    Eigen::Vector3d Jw_i = Jw.col(i);
```

Comment:
You are right. Good suggestion. I initially tried to do it without extra variables but there were some errors so I wrote this way to avoid those. I will try it again and see if I can shorten them.